<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>

<body>
  <script>
    // ————————————————————————————————————————————————————
    // const first = () =>
    //   new Promise((resolve, reject) => {
    //     console.log(3); // 同步1 3
    //     let p = new Promise((resolve, reject) => {
    //       console.log(7); //同步2 7
    //       setTimeout(() => {
    //         console.log(5); // 异步1 5
    //         resolve(6);
    //         console.log(p); // 异步2 fulfilled 1
    //       }, 0);
    //       resolve(1);
    //     });
    //     resolve(2);
    //     p.then(arg => {
    //       console.log(arg); // 微任务1 1
    //     });
    //   });

    // first().then(arg => {
    //   console.log(arg); //微任务2 2
    // });

    // console.log(4); //同步3 4
    // ————————————————————————————————————————————————————

    // let a;
    // const b = new Promise((resolve, reject) => {
    //   console.log("promise1"); // 同步1 promise1
    //   resolve();
    // }).then(() => {
    //   console.log("promise2"); // 微1 promise2
    // }).then(() => {
    //   console.log("promise3") // 微2 promise3
    // }).then(() => {
    //   console.log("promise4") // 微3 promise4
    // });
    // a = new Promise(async (resolve, reject) => {
    //   console.log(a); // 同步2 undefined
    //   console.log(b); // Promise fulfilled
    //   await b; // 等待b执行结束后，代码继续往下执行
    //   console.log(a); // 在等待b的时候 new Promise已经完成，promise pending
    //   console.log("after1"); // after1
    //   await a; // 等待a的处理结果，a 的结果在await之后，所以会一直等待
    //   resolve(true);
    //   console.log("after2");
    // });
    // console.log("end"); // 同步3 end

    // ————————————————————————————————————————————————————
    // function runAsync(x) {
    //   const p = new Promise(resolve => {
    //     setTimeout(() => {
    //       resolve(x, console.log(x));
    //     }, 1000);
    //   })
    //   return p;
    // }

    // function runReject(x) {
    //   const p = new Promise((res, rej) => {
    //     setTimeout(() => {
    //       rej(`Error:${x}`, console.log(x))
    //     }, 1000 * x);
    //   });
    //   return p;
    // }

    // // 所有的Promise都成功了，res才会打印成功的值
    // Promise.all([runAsync(1), runReject(4), runAsync(3), runReject(2)])
    //   // 13 一起出现，因为时间一样，先2后4，2的错误先到，打印错误2
    //   .then(res => console.log(res))
    //   .catch(err => console.log(err));
    // ————————————————————————————————————————————————————
    // function runAsync(x) {
    //   const p = new Promise(r => {
    //     setTimeout(() => r(x, console.log(x)), 1000)
    //   });
    //   return p;
    // }
    // // 123一起执行，启动三个定时器，123都会打印，1最先返回成功
    // Promise.race([runAsync(1), runAsync(2), runAsync(3)])
    //   // res的结果，接收返回最快的值
    //   .then(res => console.log("result:", res))
    //   .catch(err => console.log(err));
    // ————————————————————————————————————————————————————
    // async function async1() {
    //   // async没有await，剩下的代码都是同步代码
    //   console.log("1"); // 同步1
    //   // promise 的参数是个函数，这个函数是同步执行的
    //   new Promise(resolve => {
    //     // 直接执行
    //     resolve();
    //     console.log("2"); // 同步 2
    //   });
    //   console.log("3"); // 同步3
    // }
    // async1();
    // console.log("4"); // 同步4
    // ————————————————————————————————————————————————————
    // async function async1() {
    //   console.log("1"); // 同步1
    //   await async2(); // 等待 async2 执行完，包括里面的异步
    //   console.log("2"); // 微任务1
    // }
    // async function async2() {
    //   setTimeout(() => {
    //     console.log("3"); // 异步 1 
    //   }, 0);
    //   console.log("4"); // 同步2
    // }
    // async1();
    // console.log(5); // 同步4
    // ————————————————————————————————————————————————————

    // async function async1() {
    //   console.log(1); // 同步1
    //   await async2(); // 等待async2()
    //   /* 微任务区域 */
    //   console.log(2); // async2了就执行
    //   setTimeout(() => {
    //     console.log(3); // 异步3
    //   }, 0);
    //   /* 微任务区域 */
    // }
    // async function async2() {
    //   // 最先注册的计时器
    //   setTimeout(() => {
    //     console.log(4); // 异步1
    //   }, 0);
    //   console.log(5); // 同步2
    // }

    // async1(); // 从这里开始执行
    // // async1();卡住了就继续执行当前宏任务
    // setTimeout(() => {
    //   console.log(6); // 异步2
    // }, 0);
    // console.log(7); // 同步3

    // ————————————————————————————————————————————————————

    // async function async1() {
    //   console.log("1"); // 同步 2
    //   /* 这里的 Promise 是 pending 状态所以下面的代码不回执行 */
    //   await new Promise(resolve => {
    //     console.log("2"); // 同步3 因为是在括号里的
    //   });

    //   /* ↓是微任务 */
    //   console.log("3");
    //   return "4";
    // }
    // console.log("5"); // 同步1
    // async1().then(res => console.log(res));
    // console.log("6"); // 同步 4

    // ————————————————————————————————————————————————————

    // async function async1() {
    //   console.log("1"); // 同步2
    //   await new Promise(resolve => {
    //     console.log("2"); // 同步3
    //     resolve("3");
    //     // 继续按照代码顺序执行
    //   }).then(res => console.log(res)); // 微1
    //   console.log("4"); // 微2
    //   return "5";
    // }
    // console.log("6"); // 同步1
    // async1().then(res => console.log(res)); // 微3
    // console.log("7"); // 同步4


    // ————————————————————————————————————————————————————
    // async function async1() {
    //   console.log(1); // 同步2
    //   await new Promise(resolve => {
    //     console.log(2); // 同步3
    //     resolve(3);
    //   });
    //   // 靠前面的微任务先执行
    //   console.log(4); // 微1
    //   return 5;
    // }
    // console.log(6); // 同步1
    // // 接收resolve（3）
    // async1().then(res => {
    //   console.log(res); // 微2
    // });
    // new Promise(resolve => {
    //   console.log(7); // 同步 4
    //   setTimeout(() => {
    //     console.log(8); // 异步1
    //   }, 0);
    // })
    // ————————————————————————————————————————————————————
    // async function testSomething() {
    //   console.log("1"); // 同步 2
    //   return "2" // 返回的是包装后的promise 属于微任务
    // }

    // async function testAsync() {
    //   console.log("3"); // 微任务 的同步任务1
    //   return Promise.resolve("4");
    // }

    // async function test() {
    //   console.log("5"); // 同步1 
    //   // v1第一个先注册的微任务回调
    //   const v1 = await testSomething();
    //   console.log(v1); // 微1
    //   // v2 要等待 v1 的结果，v2 任务在v1后执行
    //   const v2 = await testAsync();
    //   console.log(v2); // 微3 
    //   console.log(v1, v2); // 微4
    // }
    // test();
    // let promise = new Promise(resolve => {
    //   console.log("6"); // 同步 3
    //   resolve("7") // 微任务 2
    // });
    // // 第二个注册的微任务
    // promise.then(val => console.log(val)); // 微2 结果
    // console.log("8"); // 同步 4

    // ————————————————————————————————————————————————————

    // async function async1() {
    //   await async2(); // 接收到reject，不继续执行
    //   console.log("1");
    //   return "2";
    // }

    // async function async2() {
    //   return new Promise((resolve, reject) => {
    //     console.log("3"); // 同步1
    //     reject("4")
    //   });
    // }
    // // 成功之后才会执行，所以会报错
    // async1().then(res => console.log(res)).catch(err => {
    //   console.log(err); // 捕获错误4
    // });
    // ————————————————————————————————————————————————————

    // async function async1() {
    //   try {
    //     await Promise.reject("1");
    //   } catch (e) {
    //     console.log(2); // 微1 捕获错误继续执行
    //   }
    //   console.log(3); // 2
    //   return Promise.resolve("4");
    // }

    // async1().then(res => console.log(res)); // 3
    // console.log("5"); // 同步1

    // ————————————————————————————————————————————————————

    // const async1 = async () => {
    //   console.log("1"); // 同步2

    //   setTimeout(() => {
    //     console.log("2"); // 异步2
    //   }, 2000);
    //   await new Promise(resolve => {
    //     console.log(3); // 同步3
    //   });
    //   // 微任务无状态，后续代码不执行
    //   console.log(4);
    //   return 5;
    // };

    // console.log("6"); // 同步1
    // async1().then(res => console.log(res)); // 一直等待Promise的值
    // console.log(7); // 同步 4

    // // Promise
    // Promise.resolve(1)
    //   // .then必须接收的是函数，忽略
    //   .then(2)
    //   // 忽略
    //   .then(Promise.resolve(3))
    //   // 忽略
    //   .catch(4)
    //   .then(res => console.log(res)); // 微1 1

    // setTimeout(() => {
    //   console.log(8); // 异步 1
    // }, 1000);

    // ————————————————————————————————————————————————————

    // const p1 = new Promise(resolve => {
    //     setTimeout(() => {
    //       resolve(1);
    //       console.log(2); // 3 2
    //     }, 0);
    //     resolve(3)
    //     resolve(4)
    //   })
    //   .then(res => {
    //     console.log(res); // 1  3
    //     setTimeout(() => {
    //       // console.log(p1); //4 Promise fulfilled
    //       p1.then(res => console.log(res));
    //     }, 1000);
    //     return 3
    //   })
    //   .finally(res => {
    //     // 因为前一个.then没有resolved 或 rejected 返回 undefined
    //     console.log("finally", res); // 2 finally undefined
    //   });
    
    // ————————————————————————————————————————————————————

    // const p1 = new Promise((resolve, reject) => {
    //   resolve(3);
    // })

    // p1.then((res) => {
    //   console.log(res);
    //   setTimeout(() => {
    //     p1.then(res => console.log(res))
    //   }, 1000)
    // })

    // ————————————————————————————————————————————————————
  </script>
</body>

</html>